src/daft-plan/src/optimization/rules/push_down_projection.rs

```diff
@@ -0,0 +1,425 @@
+use std::{collections::HashSet, sync::Arc};
+
+use common_error::DaftResult;
+
+use daft_core::schema::Schema;
+use daft_dsl::Expr;
+
+use crate::{
+    ops::{Aggregate, Project, Source},
+    LogicalPlan,
+};
+
+use super::{ApplyOrder, OptimizerRule, Transformed};
+
+#[derive(Default)]
+pub struct PushDownProjection {}
+
+impl PushDownProjection {
+    pub fn new() -> Self {
+        Self {}
+    }
+
+    fn try_optimize_project(
+        &self,
+        projection: &Project,
+        plan: Arc<LogicalPlan>,
+    ) -> DaftResult<Transformed<Arc<LogicalPlan>>> {
+        let upstream_plan = &projection.input;
+        let upstream_schema = upstream_plan.schema();
+
+        // First, drop this projection if it is a no-op.
+        let projection_is_noop = {
+            let maybe_column_names = projection
+                .projection
+                .iter()
+                .map(|e| match e {
+                    Expr::Column(colname) => Some(colname.clone()),
+                    _ => None,
+                })
+                .collect::<Option<Vec<_>>>();
+
+            let all_columns_equal = maybe_column_names.map_or(false, |column_names| {
+                column_names
+                    .iter()
+                    .zip(upstream_schema.names().iter())
+                    .all(|(scol, pcol)| scol.as_ref() == pcol)
+            });
+
+            let no_unused_columns = upstream_schema.names().len() == projection.projection.len();
+
+            all_columns_equal && no_unused_columns
```

Comment:
Nit: An efficiency/simplification TODO - if we restructure this a bit, we could properly short-circuit here to elide column iteration and we can avoid the allocation of the `maybe_column_names` and cloning of the column names. E.g.:
```rust
let projection_is_noop = {
    upstream_schema.names().len() == projection.projection.len() &&
    projection.projection
        .iter()
        .zip(upstream_schema.names().iter())
        .all(|(p_expr, col)| match p_expr {
            Expr::Column(p_col) => p_col == col,
            _ => False,
        })
}
```